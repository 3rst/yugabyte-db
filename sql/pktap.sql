\set ECHO

--
-- Tests for pgTAP.
--
--
-- $Id$

-- Format the output for nice TAP.
\pset format unaligned
\pset tuples_only true
\pset pager

-- Create plpgsql if it's not already there.
SET client_min_messages = fatal;
CREATE LANGUAGE plpgsql;

-- Keep things quiet.
SET client_min_messages = warning;

-- Revert all changes on failure.
\set ON_ERROR_ROLBACK 1
\set ON_ERROR_STOP true

-- Load the TAP functions.
BEGIN;
\i pgtap.sql

-- ## SET search_path TO TAPSCHEMA,public;

-- Set the test plan.
SELECT plan(26);

-- This will be rolled back. :-)
CREATE TABLE sometab(
    id    INT NOT NULL PRIMARY KEY,
    name  TEXT DEFAULT '',
    numb  NUMERIC(10, 2),
    myint NUMERIC(8)
);

/****************************************************************************/
-- Test has_pk().

\echo ok 1 - test has_pk( schema, table, description )
SELECT is(
    has_pk( 'public', 'sometab', 'public.sometab should have a pk' ),
    'ok 1 - public.sometab should have a pk',
    'has_pk( schema, table, description ) should work'
);

\echo ok 3 - test has_pk( table, description )
SELECT is(
    has_pk( 'sometab', 'sometab should have a pk' ),
    'ok 3 - sometab should have a pk',
    'has_pk( table, description ) should work'
);

\echo ok 5 - test has_pk( table )
SELECT is(
    has_pk( 'sometab' ),
    'ok 5 - Table sometab should have a primary key',
    'has_pk( table ) should work'
);

\echo ok 7 - test has_pk( schema, table, description ) fail
SELECT is(
    has_pk( 'pg_catalog', 'pg_class', 'pg_catalog.pg_class should have a pk' ),
    E'not ok 7 - pg_catalog.pg_class should have a pk\n# Failed test 7: "pg_catalog.pg_class should have a pk"',
    'has_pk( schema, table, description ) should fail properly'
);

\echo ok 9 - test has_pk( table, description ) fail
SELECT is(
    has_pk( 'pg_class', 'pg_class should have a pk' ),
    E'not ok 9 - pg_class should have a pk\n# Failed test 9: "pg_class should have a pk"',
    'has_pk( table, description ) should fail properly'
);
UPDATE __tresults__ SET ok = true, aok = true WHERE numb IN( 7, 9 );

/****************************************************************************/
-- Test col_is_pk().

\echo ok 11 - test col_is_pk( schema, table, column, description )
SELECT is(
    col_is_pk( 'public', 'sometab', 'id', 'public.sometab.id should be a pk' ),
    'ok 11 - public.sometab.id should be a pk',
    'col_is_pk( schema, table, column, description ) should work'
);

\echo ok 13 - test col_is_pk( table, column, description )
SELECT is(
    col_is_pk( 'sometab', 'id', 'sometab.id should be a pk' ),
    'ok 13 - sometab.id should be a pk',
    'col_is_pk( table, column, description ) should work'
);

\echo ok 15 - test col_is_pk( table, column )
SELECT is(
    col_is_pk( 'sometab', 'id' ),
    'ok 15 - Column sometab.id should be a primary key',
    'col_is_pk( table, column ) should work'
);

\echo ok 17 - test col_is_pk( schema, table, column, description ) fail
SELECT is(
    col_is_pk( 'public', 'sometab', 'name', 'public.sometab.name should be a pk' ),
    E'not ok 17 - public.sometab.name should be a pk\n# Failed test 17: "public.sometab.name should be a pk"\n#         have: {id}\n#         want: {name}',
    'col_is_pk( schema, table, column, description ) should fail properly'
);

\echo ok 19 - test col_is_pk( table, column, description ) fail
SELECT is(
    col_is_pk( 'sometab', 'name', 'sometab.name should be a pk' ),
    E'not ok 19 - sometab.name should be a pk\n# Failed test 19: "sometab.name should be a pk"\n#         have: {id}\n#         want: {name}',
    'col_is_pk( table, column, description ) should fail properly'
);
UPDATE __tresults__ SET ok = true, aok = true WHERE numb IN( 17, 19 );

/****************************************************************************/
-- Test col_is_pk() with an array of columns.

CREATE TABLE argh (id int not null, name text not null, primary key (id, name));

\echo ok 21 - test col_is_pk( schema, table, column[], description )
SELECT is(
    col_is_pk( 'public', 'argh', ARRAY['id', 'name'], 'id + name should be a pk' ),
    'ok 21 - id + name should be a pk',
    'col_is_pk( schema, table, column[], description ) should work'
);

\echo ok 23 - test col_is_pk( table, column[], description )
SELECT is(
    col_is_pk( 'argh', ARRAY['id', 'name'], 'id + name should be a pk' ),
    'ok 23 - id + name should be a pk',
    'col_is_pk( table, column[], description ) should work'
);

\echo ok 25 - test col_is_pk( table, column[], description )
SELECT is(
    col_is_pk( 'argh', ARRAY['id', 'name'] ),
    'ok 25 - Column argh.{id,name} should be a primary key',
    'col_is_pk( table, column[] ) should work'
);

/****************************************************************************/
-- Finish the tests and clean up.
SELECT * FROM finish();
ROLLBACK;
