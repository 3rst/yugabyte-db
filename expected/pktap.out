\unset ECHO
1..54
ok 1 - has_pk( schema, table, description ) should pass
ok 2 - has_pk( schema, table, description ) should have the proper description
ok 3 - has_pk( schema, table, description ) should have the proper diagnostics
ok 4 - has_pk( table, description ) should pass
ok 5 - has_pk( table, description ) should have the proper description
ok 6 - has_pk( table, description ) should have the proper diagnostics
ok 7 - has_pk( table ) should pass
ok 8 - has_pk( table ) should have the proper description
ok 9 - has_pk( table ) should have the proper diagnostics
ok 10 - has_pk( schema, table, description ) fail should fail
ok 11 - has_pk( schema, table, description ) fail should have the proper description
ok 12 - has_pk( schema, table, description ) fail should have the proper diagnostics
ok 13 - has_pk( table, description ) fail should fail
ok 14 - has_pk( table, description ) fail should have the proper description
ok 15 - has_pk( table, description ) fail should have the proper diagnostics
ok 16 - hasnt_pk( schema, table, description ) should fail
ok 17 - hasnt_pk( schema, table, description ) should have the proper description
ok 18 - hasnt_pk( schema, table, description ) should have the proper diagnostics
ok 19 - hasnt_pk( table, description ) should fail
ok 20 - hasnt_pk( table, description ) should have the proper description
ok 21 - hasnt_pk( table, description ) should have the proper diagnostics
ok 22 - hasnt_pk( table ) should fail
ok 23 - hasnt_pk( table ) should have the proper description
ok 24 - hasnt_pk( table ) should have the proper diagnostics
ok 25 - hasnt_pk( schema, table, description ) pass should pass
ok 26 - hasnt_pk( schema, table, description ) pass should have the proper description
ok 27 - hasnt_pk( schema, table, description ) pass should have the proper diagnostics
ok 28 - hasnt_pk( table, description ) pass should pass
ok 29 - hasnt_pk( table, description ) pass should have the proper description
ok 30 - hasnt_pk( table, description ) pass should have the proper diagnostics
ok 31 - col_is_pk( schema, table, column, description ) should pass
ok 32 - col_is_pk( schema, table, column, description ) should have the proper description
ok 33 - col_is_pk( schema, table, column, description ) should have the proper diagnostics
ok 34 - col_is_pk( table, column, description ) should pass
ok 35 - col_is_pk( table, column, description ) should have the proper description
ok 36 - col_is_pk( table, column, description ) should have the proper diagnostics
ok 37 - col_is_pk( table, column ) should pass
ok 38 - col_is_pk( table, column ) should have the proper description
ok 39 - col_is_pk( table, column ) should have the proper diagnostics
ok 40 - col_is_pk( schema, table, column, description ) fail should fail
ok 41 - col_is_pk( schema, table, column, description ) fail should have the proper description
ok 42 - col_is_pk( schema, table, column, description ) fail should have the proper diagnostics
ok 43 - col_is_pk( table, column, description ) fail should fail
ok 44 - col_is_pk( table, column, description ) fail should have the proper description
ok 45 - col_is_pk( table, column, description ) fail should have the proper diagnostics
ok 46 - col_is_pk( schema, table, column[], description ) should pass
ok 47 - col_is_pk( schema, table, column[], description ) should have the proper description
ok 48 - col_is_pk( schema, table, column[], description ) should have the proper diagnostics
ok 49 - col_is_pk( table, column[], description ) should pass
ok 50 - col_is_pk( table, column[], description ) should have the proper description
ok 51 - col_is_pk( table, column[], description ) should have the proper diagnostics
ok 52 - col_is_pk( table, column[] ) should pass
ok 53 - col_is_pk( table, column[] ) should have the proper description
ok 54 - col_is_pk( table, column[] ) should have the proper diagnostics
